pkg/scheduler/util/assumecache/assume_cache.go

```diff
@@ -0,0 +1,420 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package assumecache
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+
+	"k8s.io/klog/v2"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/tools/cache"
+)
+
+// AssumeCache is a cache on top of the informer that allows for updating
+// objects outside of informer events and also restoring the informer
+// cache's version of the object. Objects are assumed to be
+// Kubernetes API objects that are supported by [meta.Accessor].
+//
+// Objects can referenced via their key, with [cache.MetaNamespaceKeyFunc]
+// as key function.
+type AssumeCache interface {
+	// Assume updates the object in-memory only.
+	//
+	// The version of the object must be greater or equal to
+	// the current object, otherwise an error is returned.
+	//
+	// Storing an object with the same version is supported
+	// by the assume cache, but suffers from a race: if an
+	// update is received via the informer while such an
+	// object is assumed, it gets dropped in favor of the
+	// newer object from the apiserver.
+	//
+	// Only assuming objects that were returned by an apiserver
+	// operation (Update, Patch) is safe.
+	Assume(obj interface{}) error
+
+	// Restore the informer cache's version of the object.
+	Restore(key string)
+
+	// Get the object by its key.
+	Get(key string) (interface{}, error)
+
+	// GetAPIObj gets the informer cache's version by its key.
+	GetAPIObj(key string) (interface{}, error)
+
+	// List all the objects in the cache.
+	List(indexObj interface{}) []interface{}
+
+	// getImplementation is used internally by [AddTestObject], [UpdateTestObject], [DeleteTestObject].
+	getImplementation() *assumeCache
```

Comment:
If we return assumeCache specifically, then this method make less sense as part of the interface, because assumeCache is one implementation. Or let's reflect the type directly in Add/Update/Delete method, like `cache.(*assumeCache)`, I think we don't have other implementations right now.

Reply:
So basically eliminate the entire `AssumeCache` interface? Makes sense to me.

Reply:
I've pushed an update. The only real diff is this removal of the interface: https://github.com/kubernetes/kubernetes/compare/44df029ebdb6738bc724ba0d4d3e540c1c819405..7f54c5dfec3e95a2751156a3a1ae4407be68297a

But I also reorganized the commits as suggested in https://github.com/kubernetes/kubernetes/pull/124102#pullrequestreview-2021357405, so the second and third commit show the changes on top of the original assume_cache.go.

Reply:
Sorry I just mean get rid of `getImplementation()`, actually I think better to keep the interface for reference, but if @jsafrane LGTM, I'm also LGTM. It's an internal interface for volumebinding plugin, I think users wouldn't use this. Put my approval here just not block this pr merge.
/approve
/hold for @jsafrane

Reply:
I don't have a strong preference myself. Keeping the interface minimizes overall changes, but the internal method on it is ugly and the interface itself isn't needed in the first place.

@jsafrane: do you prefer this with or without the change in the last commit? I can easily drop it - I kept it separate for a reason! :smile:

Reply:
I think the interface is not really useful. There is no other implementation of it and if we need one, we can create an interface for it later.
/hold cancel
/lgtm